compiler/src/iree/compiler/DispatchCreation/SetEncoding.cpp

```diff
@@ -306,10 +306,227 @@ struct FoldFillWithSetEncoding final
+// Set padding encodings
+//===---------------------------------------------------------------------===//
+
+struct PaddedValue {
+  Value paddedValue;
+  SmallVector<Value> dynamicDims;
+};
+
+// For a given `operand`, if its producer is a `flow.dispatch.region`,
+// generate a new value for `operand` that has the padding encoding.
+// The producer `flow.dispatch.region` need not be the immediate defining op
+// of `operand`. This method tracks through operations like
+// `tensor.expand_shape/tensor.collapse_shape` to get to the producer dispatch.
+// Once the producer dispatch is found, its result is modified to be of the same
+// type as `operand` but with the padding encodings. To keep things consistent,
+// the operations that are encountered before getting to the original producing
+// `flow.dispatch.region` are replicated into the producer dispatch.
+static std::optional<PaddedValue> padProducerOfValue(RewriterBase &rewriter,
+                                                     Value operand) {
+  RankedTensorType operandType = dyn_cast<RankedTensorType>(operand.getType());
+  if (!operandType || operandType.getRank() == 0) {
+    return std::nullopt;
+  }
+
+  SmallVector<Operation *> opChain;
+  OpResult producerValue = dyn_cast<OpResult>(operand);
+  while (producerValue &&
+         !isa<IREE::Flow::DispatchRegionOp>(producerValue.getOwner())) {
+    if (!llvm::hasSingleElement(producerValue.getUses())) {
+      return std::nullopt;
+    }
+
+    // If it is an operation that we want to look past, add it to the chain
+    // and update the `producerValue`
+    Operation *currOperation = producerValue.getOwner();
+    if (isa<tensor::CollapseShapeOp, tensor::ExpandShapeOp>(currOperation)) {
+      opChain.push_back(currOperation);
+      producerValue = dyn_cast<OpResult>(currOperation->getOperand(0));
+      continue;
+    }
+
+    // Conservative, bail out.
+    return std::nullopt;
+  }
+
+  if (!producerValue) {
+    return std::nullopt;
+  }
+
+  auto producerDispatch =
+      dyn_cast<IREE::Flow::DispatchRegionOp>(producerValue.getOwner());
+  // TODO(MaheshRavishankar): Multi-result producer dispatches can be supported.
+  // Will require to move the consumer dispatch immediately after the producer
+  // instead of what is done below and move other operands of the consumer
+  // dispatch before the producer dispatch.
+  if (!producerDispatch ||
+      !llvm::hasSingleElement(producerDispatch.getBody()) ||
+      producerDispatch->getNumResults() != 1) {
+    return std::nullopt;
+  }
+  if (!llvm::hasSingleElement(producerValue.getUses())) {
+    return std::nullopt;
+  }
+
+  Location loc = producerDispatch->getLoc();
+  unsigned resultNumber = producerValue.getResultNumber();
+  // Compute the padding encoding
+  SmallVector<int32_t> paddingValue(operandType.getRank(), 0);
+  paddingValue.back() = 512;
+  auto encoding = IREE::Encoding::PadEncodingLayoutAttr::get(
+      rewriter.getContext(), paddingValue);
```

Comment:
Just a quick comment about we don't need to use PadEncodingLayoutAttr. MatmulKAttr and EncodingAttr can be specialized to PadEncodingLayoutAttr later on and the padding sizes will be resolved as well. Maybe we can introduce SetEncodingPolicy to help the control, like [ConstExprHoistingPolicy](https://github.com/iree-org/iree/blob/d33dcb57a894566f826f6a7fb2840d09d5a88053/compiler/src/iree/compiler/Dialect/Util/Analysis/Constant/ConstExpr.h#L193-L261). I'll think about it and circle back to you next week.

(I've been wanting to move this pass to a walk-based pass, but I did not find a need. Your PR is a good case shows that we don't want pattern-rewriting based driver.)

Reply:
Is this comment still valid based on offline discussion? I dont quite follow the ask here.

Reply:
No, it is not valid now. Because we agree that we can have padding strategy. Marking it resolved.